packages/apollo-server-core/src/runHttpQuery.ts

```diff
@@ -137,26 +139,37 @@ export async function runHttpQuery(
         extensions &&
         extensions.persistedQuery
       ) {
-        // It looks like we've received an Apollo Persisted Query. Apollo Server
-        // does not support persisted queries out of the box, so we should fail
-        // fast with a clear error saying that we don't support APQs. (A future
-        // version of Apollo Server may support APQs directly.)
-        throw new HttpQueryError(
-          // Return 200 to simplify processing: we want this to be intepreted by
-          // the client as data worth interpreting, not an error.
-          200,
-          JSON.stringify({
-            errors: [
-              {
-                message: 'PersistedQueryNotSupported',
-              },
-            ],
-          }),
-          true,
-          {
-            'Content-Type': 'application/json',
-          },
-        );
+        // It looks like we've received an Apollo Persisted Query.
+        const sha = extensions.persistedQuery.sha256Hash;
+
+        queryString = await optionsObject.cache.get(sha);
+        if (!queryString) {
+          throw new HttpQueryError(
+            // Return 200 to simplify processing: we want this to be intepreted by
+            // the client as data worth interpreting, not an error.
+            200,
+            JSON.stringify({
+              errors: formatApolloErrors([new PersistedQueryNotFoundError()], {
+                formatter: optionsObject.formatError,
+                debug,
+                logFunction: optionsObject.logFunction,
+              }),
+            }),
+            true,
+            {
+              'Content-Type': 'application/json',
+            },
+          );
+        }
+      } else if (extensions && extensions.persistedQuery) {
+        const sha = extensions.persistedQuery.sha256Hash;
+        const calculatedSha = sha256()
+          .update(queryString)
+          .digest('hex');
+        if (sha !== calculatedSha) {
+          throw new HttpQueryError(400, `provided sha does not match query`);
+        }
+        await optionsObject.cache.set(sha, queryString);
```

Comment:
Ditto for guard.

In engineproxy we explicitly did this write asynchronously and logged errors but didn't wait on the result.  Might be a good idea here too, as long as we're comfortable logging errors (or doing something else with them)!